### PR TITLE
Optimizer: Fix range extraction for CNF(conjunctive normal form) | tidb-test=pr/2341

### PR DESCRIPTION
**Pull request by ghazalfamilyusa**
### What problem does this PR solve?

Issue Number: Ref #41598

### Problem Summary:
Index range extraction in the optimizer can handle disjunctions like 
((a = 1 and b = 2 and c > 3) or (a = 4 and b = 5 and c > 6)) and builds the proper index range like "(1 2 3,1 2 +inf], (4 5 6,4 5 +inf]" for this example. The problem is that when we add another conjunct like d > 3 then the optimizer takes another path which find only the point ranges "[1,1], [4,4]" in this example. Finding point ranges is OK for complex CNF where we only pick ranges from only one conjunct, But, in this case, the best conjunct is also better than point ranges. 

### What changed and how does it work?
 We added code that picks the best conjunct ranges if it is more selective than the point ranges. For example, "(1 2 3,1 2 +inf], (4 5 6,4 5 +inf]" is more selective than "[1,1], [4,4]" . We implemented that by checking if the best CNF ranges is a subset of the point ranges. 

### Check List

Tests 

- [X] Unit test
- [] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by zanmato1984 — APPROVED** [review]

**Comment by ghazalfamilyusa** [comment]
cc @XuHuaiyu for expression changes

**Comment by ghazalfamilyusa** [comment]
cc @qw4990

**Comment by elsa0520** [comment]
> ### What problem does this PR solve?
> Issue Number: Ref #41598
> 
> ### Problem Summary:
> Index range extraction in the optimizer can handle disjunctions like ((a = 1 and b = 2 and c > 3) or (a = 4 and b = 5 and c > 6)) and builds the proper index range like "(1 2 3,1 2 +inf], (4 5 6,4 5 +inf]" for this example. The problem is that when we add another conjunct like d > 3 then the optimizer takes another path which find only the point ranges "[1,1], [4,4]" in this example. Finding point ranges is OK for complex CNF where we only pick ranges from only one conjunct, But, in this case, the best conjunct is also better than point ranges.
> 
> ### What changed and how does it work?
> We added code that picks the best conjunct ranges if it is more selective than the point ranges. For example, "(1 2 3,1 2 +inf], (4 5 6,4 5 +inf]" is more selective than "[1,1], [4,4]" . We implemented that by checking if the best CNF ranges is a subset of the point ranges.
> 
> ### Check List
> Tests
> 
> * [x]  Unit test
> * [] Integration test
> * [ ]  Manual test (add detailed scripts or steps below)
> * [ ]  No need to test
>   > * [ ]  I checked and no code files have been changed.
> 
> Side effects
> 
> * [ ]  Performance regression: Consumes more CPU
> * [ ]  Performance regression: Consumes more Memory
> * [ ]  Breaking backward compatibility
> 
> Documentation
> 
> * [ ]  Affects user behaviors
> * [ ]  Contains syntax changes
> * [ ]  Contains variable changes
> * [ ]  Contains experimental features
> * [ ]  Changes MySQL compatibility
> 
> ### Release note
> Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.
> 
> ```
> None
> ```

What's different between with ```and d>3``` and without ```and d>3``` ? 
Whatever with or without ``` and d>3``` , the range of DNF ```((a = 1 and b = 2 and c > 3) or (a = 4 and b = 5 and c > 6)) ``` should be  "(1 2 3,1 2 +inf], (4 5 6,4 5 +inf]".
Why the range of DNF changed to point range when there is additional predicate ```and d>3``` ?

**Comment by ghazalfamilyusa** [comment]
> What's different between with `and d>3` and without `and d>3` ? Whatever with or without ` and d>3` , the range of DNF `((a = 1 and b = 2 and c > 3) or (a = 4 and b = 5 and c > 6)) ` should be "(1 2 3,1 2 +inf], (4 5 6,4 5 +inf]". Why the range of DNF changed to point range when there is additional predicate `and d>3` ?
------------
@elsa0520 ; this is what I talked about in the design document and in the summary above. The code will take the point ranges path when you have multiple conjuncts regardless if these conjuncts are relevant or not like (d > 3).

**Review by XuHuaiyu — APPROVED** [review]
The expression change LGTM

**Comment by elsa0520** [comment]
> > What's different between with `and d>3` and without `and d>3` ? Whatever with or without ` and d>3` , the range of DNF `((a = 1 and b = 2 and c > 3) or (a = 4 and b = 5 and c > 6)) ` should be "(1 2 3,1 2 +inf], (4 5 6,4 5 +inf]". Why the range of DNF changed to point range when there is additional predicate `and d>3` ?
> 
> @elsa0520 ; this is what I talked about in the design document and in the summary above. The code will take the point ranges path when you have multiple conjuncts regardless if these conjuncts are relevant or not like (d > 3).

But in the design we will deal with the CNF intersection. So after you support Computing intersection for CNF . We can merge the ((a = 1 and b = 2 and c > 3) or (a = 4 and b = 5 and c > 6)) and d>3 . Then we don't need to support PR alone, isn't it?

**Comment by ghazalfamilyusa** [comment]
> > > What's different between with `and d>3` and without `and d>3` ? Whatever with or without ` and d>3` , the range of DNF `((a = 1 and b = 2 and c > 3) or (a = 4 and b = 5 and c > 6)) ` should be "(1 2 3,1 2 +inf], (4 5 6,4 5 +inf]". Why the range of DNF changed to point range when there is additional predicate `and d>3` ?
> > 
> > 
> > @elsa0520 ; this is what I talked about in the design document and in the summary above. The code will take the point ranges path when you have multiple conjuncts regardless if these conjuncts are relevant or not like (d > 3).
> 
> But in the design we will deal with the CNF intersection. So after you support Computing intersection for CNF . We can merge the ((a = 1 and b = 2 and c > 3) or (a = 4 and b = 5 and c > 6)) and d>3 . Then we don't need to support PR alone, isn't it?

@elsa0520 : seems my previous reply is not clear. The intersection is not 100% guaranteed at least for a while. Initially, we will have a flag and later on we may implement a heuristic solution. So, we need the point ranges in case the flag is off or if the intersection is not done for all the conjuncts.  Now, with the reality of having point ranges (just in case) we still need to choose bestCNF over range. we can remove point ranges support and related PRs once we have: 100% intersection applied with no flag.  Hope this makes sense and if not we can chat

File: pkg/util/ranger/detacher.go [hunk 1]
```diff
@@ -462,6 +463,18 @@ func (d *rangeDetacher) detachCNFCondAndBuildRangeForIndex(conditions []expressi
 		// TODO: we will optimize it later.
 		res.RemainedConds = AppendConditionsIfNotExist(res.RemainedConds, remainedConds)
 		res.Ranges = ranges
+		if bestCNFItemRes != nil && res != nil && len(res.Ranges) != 0 {
```
Comment (elsa0520) [inline pkg/util/ranger/detacher.go:466]: Could you please comment this case " (heuristics applied for long lists or we turn off the intersection)" in front of here?

Comment (ghazalfamilyusa) [reply pkg/util/ranger/detacher.go:469]: done

**Review by elsa0520 — APPROVED** [review]
LGTM

**Pull request merged**